### PR TITLE
FEAT: Eggnog formats and types

### PR DESCRIPTION
**Pull request by Keegan-Evans**
Clearly not done, but @misialq, I wanted you to be able to see where I was getting with these. I tweaked the formats to be more in line with the most recent document, and tweaked/fixed some of the basic tests. Monday I am going to work on validation for NOG and hopefully get rolling on the OG and KEGG stuff.

### CODE REVIEW

**Comment by misialq** [comment]
Hey @Keegan-Evans, I only now realized that you tagged me for review a while ago... Do you want me to go through the code or are you still working on it? (I _glanced_ through it and so far it seems 👍)

File: q2_types_genomics/eggnog/_format.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+from ..plugin_setup import plugin
+
+
+class ArbitraryHeaderTSVFmt(model.TextFileFormat):
+    def _check_seperator(self, level):
+        with self.open() as fh:
+            for i, line in enumerate(fh, 1):
+                if (not re.search(r'\t', line) and
+                        i >= self.header + 1 and i < self.footer):
```
Comment (misialq) [inline q2_types_genomics/eggnog/_format.py:23]:
```suggestion
                if (not re.search(r'\t', line) and
                        self.header + 1 <= i < self.footer):
```

File: q2_types_genomics/eggnog/_validator.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+def validate_og(data: pd.DataFrame, level):
+    _check_fields(data=data, level=level, reference_fields=og_fields)
+
+
+@plugin.register_validator(FeatureData[KEGG])
+def validate_kegg(data: pd.DataFrame, level):
+    _check_fields(data=data, level=level, reference_fields=kegg_fields)
+
+
+def _check_fields(data: pd.DataFrame, level, reference_fields):
```
Comment (misialq) [inline q2_types_genomics/eggnog/_validator.py:51]: Does this function need the `level` arg? Unless you anticipate some usage in the near future, please remove.

File: q2_types_genomics/eggnog/_validator.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                  'BiGG_Reaction', 'PFAMs',
+                  ])
+
+og_fields = set(['eggNOG', 'OGs', 'narr_og_name', 'narr_og_cat',
+                 'nar_og_desc',
+                 ])
+
+kegg_fields = set(['KEGG_ko', 'KEGG_Pathway', 'KEGG_Module', 'KEGG_Reaction',
+                   'KEGG_rclass', 'BRITE', 'KEGG_TC',
+                   ])
```
Comment (misialq) [inline q2_types_genomics/eggnog/_validator.py:33]: Please replace with a set literal (here and above):
```suggestion
kegg_fields = {'KEGG_ko', 'KEGG_Pathway', 'KEGG_Module', 'KEGG_Reaction',
               'KEGG_rclass', 'BRITE', 'KEGG_TC'}
```

**Review by misialq — CHANGES_REQUESTED** [review]
Hey @Keegan-Evans, thanks for those new formats! Two high-level comments:
- Could you please rename the test data files using snake case? It's a bit difficult to read the names.
- Please update your branch with the newest `main` to re-run the CI with coverage.

Otherwise, please see some comments inline.

Thanks!

File: q2_types_genomics/__init__.py [hunk 4]
```diff
@@ -15,3 +15,4 @@
 
 importlib.import_module('q2_types_genomics.per_sample_data')
 importlib.import_module('q2_types_genomics.genome_data')
+importlib.import_module('q2_types_genomics.eggnog')
```
[the remark this answers is not in the document]
Comment (Keegan-Evans) [reply q2_types_genomics/__init__.py:18]: Nope, just put there to keep contained away from other changes(you know in addition to being in a separate branch 😅) I will get them moved over.

File: q2_types_genomics/eggnog/_util.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+def parse_footer_line(filepath, comment_char='#'):
+    cnt = 0
+    past_header = False
+    with filepath.open() as fh:
+        while True:
+            line = fh.readline()
+            cnt += 1
+            if not past_header and not line.startswith(comment_char):
+                print(line)
```
[the remark this answers is not in the document]
Comment (Keegan-Evans) [reply q2_types_genomics/eggnog/_util.py:50]: Yep 😬

File: q2_types_genomics/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Keegan-Evans) [reply q2_types_genomics/__init__.py]: https://github.com/Keegan-Evans/q2-types-genomics/commit/1164b485179fff6d2e08ae9f559b8ccd9df89c7d

File: q2_types_genomics/eggnog/tests/test_format.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+from q2_types_genomics.eggnog import ArbitraryHeaderTSVFmt
+
+from qiime2.plugin.testing import TestPluginBase
+
+
+class TestArbitraryHeaderTSVFmt(TestPluginBase):
+    """This format is for files written as TSVs with arbitrary header and/or
+    footer lengths and locations, verification of content should be performed
+    using Semantic Validators"""
```
[the remark this answers is not in the document]
Comment (Keegan-Evans) [reply q2_types_genomics/eggnog/tests/test_format.py]: moved.

File: q2_types_genomics/eggnog/_format.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                    raise ValueError("No correct separator detected in input "
```
[the remark this answers is not in the document]
Comment (Keegan-Evans) [reply q2_types_genomics/eggnog/_format.py]: Indeed it should 😅

File: q2_types_genomics/eggnog/tests/test_format.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        has_run = False
+        filename = 'noheadersample.txt'
+        filepath = self.get_data_path(filename)
+
+        fmt = ArbitraryHeaderTSVFmt(filepath, mode='r')
+        fmt.validate()
+        has_run = True
+        assert has_run
+
+        print(fmt.header)
```
[the remark this answers is not in the document]
Comment (Keegan-Evans) [reply q2_types_genomics/eggnog/tests/test_format.py]: yep, removed

File: q2_types_genomics/eggnog/tests/test_validators.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    def test_raise_on_missing_nog_field(self):
+        has_run = False
+        filename = "missingnogannotations.txt"
+        filepath = self.get_data_path(filename)
+        bad_df = pd.read_csv(filepath, sep='\t', header=4)
+
+        with self.assertRaisesRegex(
+                ValidationError,
+                r".*Required fields not found in data.*"):
+            has_run = True
```
[the remark this answers is not in the document]
Comment (Keegan-Evans) [reply q2_types_genomics/eggnog/tests/test_validators.py]: Yes, but if you put them after running the validation, when the error is raised, then it does not actually end up running the assignment statement 😵‍💫

File: q2_types_genomics/eggnog/tests/test_validators.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+    # og validators
+    def test_og_fields_passing(self):
+        has_run = False
+        filename = "ogannotations.txt"
+        filepath = self.get_data_path(filename)
+        good_df = pd.read_csv(filepath, sep='\t', header=4)
+        if good_df is not None:
+            has_run = True
+        assert has_run
```
[the remark this answers is not in the document]
Comment (Keegan-Evans) [reply q2_types_genomics/eggnog/tests/test_validators.py]: I think I have re-written all of these, I was not sure how I wanted to actually run the validators, as I wanted to make sure they were being run by the framework during view access, I think I have it resolved now.

**Pull request closed without merging**